IPsec VinE Implementation

ldelossa opened the pull request:
Implement IPsec VinE

See commits for details. 

```release-note
Add support for VXLAN in IPsec (VinE)
```

pchaigno committed:
daemon: Reload bpf_host first in case of IPsec upgrade

This commit was originally applied in ca9c056deb31f6e0747c951be24b25d67ea99d6d
removed in 152867f906715c926d7fe5d65812e13735b8c53c and is now being
reapplied to support moving IPsec hooks to the egress device.

To support upgrades properly we ensure that `bpf_host` is installed
first when IPsec is enabled.

This ensures the transition to `bpf_lxc` programs which no longer
perform IPsec encryption do not result in leaked traffic. The IPsec
enabled `bpf_host` programs will be in place prior to the IPsec disabled
`bpf_lxc` transition.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

ldelossa committed:
ipsec,config: require devices when IPsec is enabled

To fully support Encrypted Overlay and moving IPsec encrypt hooks to the
egress device we must ensure the `to-netdev` program exists on native
devices.

Update the `DaemonConfig.AreDevicesRequired` function to return `true`
when IPsec is enabled.

This is a more general option which encompasses the `EnableIPSecEncryptedOverlay`
flag.

Moving forward Encrypted Overlay will be the default and this flag will
be removed.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bpf,ipsec: move encryption hooks to egress device

Move IPsec encryption hooks to the native egress device.

Prior to this commit IPsec hooks would occur at pod egress (or at proxy
egress if l7 policy is enforced).

This move accomplish two goals:
1. Consolidate the encryption datapath flow between IPsec and Wireguard
2. Encrypt security IDs when tunnel mode is enabled (VinE/Encrypted
   Overlay)

Thus, when this commit lands IPsec hooks will begin to be applied at
bpf_host and only after the vlxan driver has encapsulated a packet, if
tunnel mode is enabled.

To cleanly migrate to the new IPsec datapath the Cilium cluster MUST be
patched to the latest v1.17 patch release and then upgraded to v1.18.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
ipsec: update unit tests for VinE support

Remove the tests which are no longer valid for IPsec.

This includes the group of tests which focused on bpf_lxc and encrypted
overlay.

IPsec encryption hooks no longer run at bpf_lxc and encrypted overlay is
deprecated.

Add two new tests:
1. Integration test with bpf_host which ensures the IPsec hooks are
   called for traffic about to leave the host
2. Unit tests for the IPsec encryption hook and its hairpin
   functionality.

The latter ensures the `ctx` is prepared correctly for the hairpin and
a call to ctx_redirect is made to recirculate the packet into the stack
for XFRM processing

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
cli: enable v2 pod-to-pod encryption tests for IPsec

Enable the v2 pod-to-pod encryption tests for IPsec and not just
wireguard, now that VinE is implemented.

Additionally, we can now disable the v1 pod-to-pod encryption tests for
all encryption modes, since IPsec no longer requires it to run.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -881,11 +876,6 @@ handle_ipv4_cont(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
-	if (from_proxy && info->tunnel_endpoint && encrypt_key)
-		return set_ipsec_encrypt(ctx, encrypt_key, info->tunnel_endpoint,
-					 info->sec_identity, true, false);
-
```
These packets will then trickle down into `to-netdev` instead, and get encrypted there. Sounds good.

Note that this requires that the BPF programs on native devices are re-loaded first.

ldelossa replied:
Yup, I'm about to add a commit which re-implements Paul's original change that forces host device attachment first.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1697,6 +1676,39 @@ int cil_to_host(struct __ctx_buff *ctx)
+	 * exists which will mark the packet with 0x200 and trigger a local
+	 * delivery as part of L7 Proxy TPROXY mechanism.
+	 *
+	 * This iptables rule is ignored by the mark MARK_MAGIC_PROXY_TO_WORLD.
```
This is `inboundProxyRedirectRule()`.

ldelossa replied:
Cool. I'll update the comment to be more specific and callout that function.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1697,6 +1676,39 @@ int cil_to_host(struct __ctx_buff *ctx)
+	 * and allow it to enter the foward path once punted to stack.
+	 */
+	if (ctx->mark == 0)
+		ctx->mark |= MARK_MAGIC_PROXY_TO_WORLD;
```
We'll need to give `MARK_MAGIC_PROXY_TO_WORLD` a better name then (fine as follow-up) :)

ldelossa replied:
Yup, naming is confusing, just didn't want to hack in "yet another mark" to both the datapath and our iptables rule.

julianwiedmann commented on bpf/lib/encrypt.h:
```diff
@@ -204,6 +205,120 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+		 * encapsulated packets get here the destination address is
+		 * already a cluster node IP.
+		 *
+		 * cluster node IPs will have 0.0.0.0 as their tunnel_endpoint
```
Could you double-check this part? I believe it changed with https://github.com/cilium/cilium/pull/36978.

ldelossa replied:
Yup, you are correct:

```
cilium bpf ipcache list | grep -E "^172.*"
172.18.0.2/32               identity=1 encryptkey=0 tunnelendpoint=0.0.0.0 flags=skiptunnel
172.18.0.3/32               identity=7 encryptkey=0 tunnelendpoint=172.18.0.3 flags=skiptunnel
172.18.0.4/32               identity=6 encryptkey=0 tunnelendpoint=172.18.0.4 flags=skiptunne
```

Let me see if this simplifies things...

ldelossa replied:
So, it does not clean up too much, since we still need to get the source sec ID from `ctx_is_overlay(ctx)` and subsequent `get_identity(ctx)`. 

I think at the most we can update to:

```
		if (ctx_is_overlay(ctx)) {
			ret = set_ipsec_encrypt(ctx, 0, dst->tunnel_endpoint,
						get_identity(ctx), false,
						true);
			if (ret != CTX_ACT_OK)
				return ret;
			goto overlay_encrypt;
		}
```

But not sure this gains us that much. 

At the least, I can update the comments.

ldelossa replied:
Hmm, actually testing the above, we run into some errors. 🤔

ldelossa replied:
Talked with @julianwiedmann out of band. Because we still need to get CTX from where overlay prog stuffs it, we will simply update the comments here, as they are incorrect and misleading.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1490,26 +1480,15 @@ int cil_to_netdev(struct __ctx_buff *ctx)
+#if defined(ENABLE_IPSEC)
+	if ((ctx->mark & MARK_MAGIC_HOST_MASK) != MARK_MAGIC_ENCRYPT) {
+		ret =  ipsec_maybe_redirect_to_encrypt(ctx, proto);
+		if (ret == CTX_ACT_REDIRECT)
```
Could we sketch out how this behaves when seeing "legacy" traffic that is still in-flight, but already encrypted?

* native routing
   * should be marked with `MARK_MAGIC_ENCRYPT`, and thus skip this path
* overlay routing
   * is marked with `MARK_MAGIC_OVERLAY`. Would we then encrypt a second time?

jschwinger233 replied:
> overlay routing: is marked with MARK_MAGIC_OVERLAY. Would we then encrypt a second time?

Interesting, maybe 1.17 is already capable of handling ESP-Vxlan-ESP traffic. I'm optimistic about this.

ldelossa replied:
I am planning on solving this issue by detecting ESP traffic when it is passed to the overlay device. 
This will also be helpful when back porting to v1.17, since we can determine if we should ESP encrypt any stale packets due to downgrade. 


```diff
diff --git a/bpf/bpf_overlay.c b/bpf/bpf_overlay.c
index 16f7a99c36..4f527abfb1 100644
--- a/bpf/bpf_overlay.c
+++ b/bpf/bpf_overlay.c
@@ -764,8 +764,12 @@ int cil_to_overlay(struct __ctx_buff *ctx)
 	if (!ctx_get_tunnel_key(ctx, &tunnel_key, TUNNEL_KEY_WITHOUT_SRC_IP, 0))
 		src_sec_identity = get_id_from_tunnel_id(tunnel_key.tunnel_id,
 							 ctx_get_protocol(ctx));
-
-	set_identity_mark(ctx, src_sec_identity, MARK_MAGIC_OVERLAY);
+#ifdef ENABLE_IPSEC
+	if (is_esp(ctx, proto))
+		set_identity_mark(ctx, src_sec_identity, MARK_MAGIC_OVERLAY_ENCRYPTED);
+	else
+#endif
+		set_identity_mark(ctx, src_sec_identity, MARK_MAGIC_OVERLAY);
 
 #ifdef ENABLE_NODEPORT
 	if (snat_done) {
diff --git a/bpf/lib/common.h b/bpf/lib/common.h
index b3cade6026..ec8369c7d2 100644
--- a/bpf/lib/common.h
+++ b/bpf/lib/common.h
@@ -707,6 +707,13 @@ enum metric_dir {
 #define MARK_MAGIC_TO_PROXY		0x0200
 #define MARK_MAGIC_SNAT_DONE		0x0300
 #define MARK_MAGIC_OVERLAY		0x0400
+/* used to indicate encrypted traffic was tunnel encapsulated
+ * this is useful in the IPsec code paths where we need to know if overlay
+ * traffic is encrypted or not.
+ *
+ * the SPI bit can be reused since this magic mark is only used POST encryption.
+ */
+#define MARK_MAGIC_OVERLAY_ENCRYPTED	0x1400
 #define MARK_MAGIC_EGW_DONE		0x0500
 
 #define MARK_MAGIC_KEY_MASK		0xFF00
diff --git a/bpf/lib/encrypt.h b/bpf/lib/encrypt.h
index 1ed28074a0..d68036b710 100644
--- a/bpf/lib/encrypt.h
+++ b/bpf/lib/encrypt.h
@@ -258,6 +258,9 @@ ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
 		 * set_ipsec_encrypt to obtain the correct node ID and spi.
 		 */
 		if (ctx_is_overlay(ctx)) {
+			/* if this is already encrypted, just pass it to stack */
+			if (ctx_is_overlay_encrypted(ctx))
+				return CTX_ACT_OK;
 			ret = set_ipsec_encrypt(ctx, 0, ip4->daddr,
 						get_identity(ctx), false,
 						true);
diff --git a/bpf/lib/overloadable_skb.h b/bpf/lib/overloadable_skb.h
index 236481d317..0eb32e6c78 100644
--- a/bpf/lib/overloadable_skb.h
+++ b/bpf/lib/overloadable_skb.h
@@ -262,6 +262,14 @@ static __always_inline bool ctx_is_overlay(const struct __sk_buff *ctx)
 	return (ctx->mark & MARK_MAGIC_HOST_MASK) == MARK_MAGIC_OVERLAY;
 }
 
+static __always_inline bool ctx_is_overlay_encrypted(const struct __sk_buff *ctx)
+{
+	if (!is_defined(HAVE_ENCAP))
+		return false;
+
+	return (ctx->mark & MARK_MAGIC_KEY_MASK) == MARK_MAGIC_OVERLAY_ENCRYPTED;
+}
+
 static __always_inline bool ctx_mark_is_wireguard(const struct __sk_buff *ctx)
 {
 	if (!is_defined(ENABLE_WIREGUARD))
```

julianwiedmann replied:
> I am planning on solving this issue by detecting ESP traffic when it is passed to the overlay device. This will also be helpful when back porting to v1.17, since we can determine if we should ESP encrypt any stale packets due to downgrade.
> 
> ```diff
> diff --git a/bpf/lib/common.h b/bpf/lib/common.h
> index b3cade6026..ec8369c7d2 100644
> --- a/bpf/lib/common.h
> +++ b/bpf/lib/common.h
> @@ -707,6 +707,13 @@ enum metric_dir {
>  #define MARK_MAGIC_TO_PROXY		0x0200
>  #define MARK_MAGIC_SNAT_DONE		0x0300
>  #define MARK_MAGIC_OVERLAY		0x0400
> +/* used to indicate encrypted traffic was tunnel encapsulated
> + * this is useful in the IPsec code paths where we need to know if overlay
> + * traffic is encrypted or not.
> + *
> + * the SPI bit can be reused since this magic mark is only used POST encryption.
> + */
> +#define MARK_MAGIC_OVERLAY_ENCRYPTED	0x1400
>  #define MARK_MAGIC_EGW_DONE		0x0500



wow, that's a lot of smart games with aliasing MARK_MAGIC_OVERLAY in (hopefully) all the right places 😁. This can go away again in v1.19, correct?

Just a nit - could you please also update the corresponding spot in pkg/datapath, so that we keep track of the mark usage?

ldelossa replied:
Lol it's a bit sneaky, but luckily, no other code path changes where its checked. We only **add** a check in the ipsec code. 

Yes, will update pkg/datapath.

ldelossa replied:
Ah forgot to answer. Yes, I think this all goes away in v1.19.

ldelossa replied:
This is implemented and works. Going to resolve.

julianwiedmann commented on bpf/lib/encrypt.h:
```diff
@@ -204,6 +205,120 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+
+	if (!ipsec_redirect_sec_id_ok(dst->sec_identity) ||
+	    !ipsec_redirect_sec_id_ok(src->sec_identity))
+		return CTX_ACT_OK;
```
I understand why we check the source's security identity here (we only encrypt pod-originating traffic), but looking at the destination's security identity is a new thing, right? Why not rely on the `spi` in the ipcache entry?

ldelossa replied:
I remember discussion of reducing the number of maps we store the SPI in. 
I think the consensus was to collapse onto the nodeID map, being that you can lookup the destination node for the pod and get both the SPI and the nodeID at once. 

However, let me refresh this convo, I need to recall if that's the case. 
If that's the case, we may just want to confirm that both our src and dst are valid for IPsec as a sanity check here and not have to get that out of `set_ipsec_mark`.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1697,6 +1676,39 @@ int cil_to_host(struct __ctx_buff *ctx)
+	 * Technically, it is also ignored by MARK_MAGIC_ENCRYPT but reusing
+	 * this mark breaks further processing as its used in the XFRM subsystem.
+	 *
+	 * Therefore, if the packet's mark is zero, indicating it was forwarded
+	 * from 'cilium_host', mark the packet with MARK_MAGIC_PROXY_TO_WORLD
```
This construction is certainly "fun" ;). My main concern would be that this applies on "both" sides of the veth pair (we attach `to-host` twice). Can we limit this to exactly where it's needed?

ldelossa replied:
Thanks for calling this out, it's been a concern of mine too. 

Essentially we only want this to run on `cilium_net`. 
We could check `THIS_INTERFACE_MAC == CILIUM_NET_MAC` check.

julianwiedmann replied:
> Essentially we only want this to run on `cilium_net`.
> We could check `THIS_INTERFACE_MAC == CILIUM_NET_MAC` check.

:+1: I think we typically use `THIS_INTERFACE_IFINDEX`, but same-same. Best double-check if the loader is already injecting those macros correctly before you dig too deep, this feels easy to miss.

ldelossa replied:
Oh, even easier to compare than, will confirm.

ldelossa replied:
This is handled in latest commit. All my local tests are passing. Lets run it thru CI.

julianwiedmann reviewed (commented):
Some comments in-line :)

jschwinger233 commented on bpf/bpf_host.c:
```diff
@@ -1697,6 +1676,39 @@ int cil_to_host(struct __ctx_buff *ctx)
+	 * this mark breaks further processing as its used in the XFRM subsystem.
+	 *
+	 * Therefore, if the packet's mark is zero, indicating it was forwarded
+	 * from 'cilium_host', mark the packet with MARK_MAGIC_PROXY_TO_WORLD
```
It's been discussed months ago and I really should have brought it up earlier, but, given wireguard routes proxied traffic directly to native devices (on native routing), what if ipsec follows the same approach? This means changing the 2005 routing table from `default dev cilium_host` to ... whatever there are in the "main" routing table, with MTU adjustments.

(actually it's unclear to me why wireguard + proxy never saw mtu issue? I should have a look at it)

then, we might not need this "smelly" code?
(well, if the `MARK_MAGIC_PROXY_TO_WORLD` can be changed to a better name, it's not that smelly)

ldelossa replied:
@jschwinger233 I do not want to get into the scenario where we are constantly trying to `mirror` "main" routing table changes into a shadow "2005" table. The Cilium code has multiple sources of routes, along with endpoint routes being a configuration. I find it a harder sell to say "mirror main into 2005" then to simply send back into `cilium_host` for me. 

But yes, we should def change the name, to not make that confusing...

jschwinger233 replied:
Agreed with 

> do not want to get into the scenario where we are constantly trying to mirror "main" routing table changes into a shadow "2005" table

otherwise it'll be chaotic when it comes to eni / dynamic routing scenaio.

ldelossa replied:
I agree, gets too complex.

jschwinger233 commented on bpf/lib/encrypt.h:
```diff
@@ -204,6 +205,120 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+
+	/* mark packet for encryption based on ipcache's key */
+	ret = set_ipsec_encrypt(ctx, dst->key, dst->tunnel_endpoint,
+				src->sec_identity, false, true);
```
`set_ipsec_encrypt(....., use_meta=false, use_spi_from_map=true)`:

when it's called with `use_spi_from_map=true`, the 2nd argument `spi=dst->key` is not used at all, right? we'll always get spi from node_map.

though this is https://github.com/cilium/cilium/issues/34790, I'm still concerned this step is too big: moving ipsec hooks to eth0 and using spi from node_map instead of ipcache in the meantime seems difficult in terms of risk control.

ldelossa replied:
@jschwinger233 so you would rather move to using node-map more, later? Not during this change?

I am not opposed to switching to using node map in a later commit. 

But it should be noted, we depend on this behavior when in tunnel mode. This was a requirement for the original Encrypted Overlay implementation. If we depend on it there, I don't see why we can't depend on it here as well.

jschwinger233 replied:
> you would rather move to using node-map more, later? Not during this change?

Yeah. I'm not opposing it, just trying to make this change noticed and reviewed by more engineers, increasing the chances of catching potential bugs.

jschwinger233 replied:
Let's use node map in a separate PR shall we? My points are:
1. Neither PR description or commit message mentioned this change
2. We expect only tunnel routing is affected by this PR
3. It's "hidden" in a big commit that engineers are easy to overlook
4. This looks like the last place to use ipcache encryptkey field for ipsec, we can also delete the obsolete control plane code once switching to node map

ldelossa replied:
@jschwinger233 Yup, lets move away from node map then. I'll work on this commit after I finish up getting the v1.17 backport branch in shape.

ldelossa replied:
Running tests for this change now, including backport branch.

ldelossa replied:
Spoke with @jschwinger233 out of band for this. 

Introducing the use of ipcache key lookup instead of node map lookup in `bpf_host` ipsec hooks introduced a new error with key rotation. 

For the sake of time, I will create an issue which makes not of this but leave the PR using the node map key so we can merge the initial VinE support.

ldelossa committed:
ipsec: detect encrypted overlay traffic via mark

Detect if traffic has been encrypted before traversing the overlay
driver via a mark.

This is useful for VinE, where encryption occurs at the ultimate egress
device. If a packet was previously encrypted by the pre-VinE bpf_lxc
programs, we can detect this and avoid performing double encryption.

Likewise, if a packet arrives at the egress device and is unencrypted
tunnel traffic, we can encrypt it to ensure the traffic does not leak.

Implementation of the previous statement will be done in a subsequent
commit.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

julianwiedmann commented on daemon/cmd/state.go:
```diff
@@ -322,7 +322,26 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
+		if ep.IsHost() && option.Config.EnableIPSec {
+			// The host endpoint was handled above.
+			continue
+		}
```
For bisect reasons, I think we want this patch earlier in the series.

(and the new `OVERLAY` mark now *also* requires that `bpf_host` is regenerate prior to `bpf_overlay` :slightly_smiling_face: )

ldelossa replied:
@julianwiedmann 

ACK on reorder of commit. 

Re: 
> and the new OVERLAY mark now also requires that bpf_host is regenerate prior to bpf_overlay 🙂

I don't think this is the case, since we will backport this change into v1.17. Therefore, `bpf_overlay` will be marking packets correctly with `MARK_MAGIC_OVERLAY_ENCRYPTED` prior to the v1.18 upgrade. Does my thinking check out there?

julianwiedmann replied:
>since we will backport this change into v1.17. Therefore, bpf_overlay will be marking packets correctly with MARK_MAGIC_OVERLAY_ENCRYPTED prior to the v1.18 upgrade. Does my thinking check out there?

ah yep, that should work! Do you have a v1.17 draft somewhere already?

ldelossa replied:
@julianwiedmann doing this right meow. Unless I run into any horrible conflicts, should be up pretty soon.

ldelossa replied:
Backport draft: https://github.com/cilium/cilium/pull/37993

@jschwinger233 identified one of the existing issues with Cluster Mesh.
The other two failing issues are more trivial and just appear to be unit tests items. 

I am working on hacking the ipsec upgrade tests in this PR to reference the v1.17 backport image in that linked PR. We'll run tests again with the backport v1.17 image and ensure all our downgrade leaks go away.

jschwinger233 committed:
gh: ci-clustermesh: Manually assert no "No node ID found" drops

This commit is added to work around an issue with the cluster mesh
testing.

This issue is being tracked and worked on saparately to this PR in
issue: 38113

See issue for full details.

Signed-off-by: gray <gray.liang@isovalent.com>

jschwinger233 commented on bpf/lib/encrypt.h:
```diff
@@ -205,6 +206,140 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+			/* see comment in the native-routing mode call. */
+			ret = set_ipsec_encrypt(ctx, 0, ip4->daddr,
+						get_identity(ctx), true,
+						true);
```
`set_ipsec_encrypt(..., use_meta=true, ...)`: do you think we can set `use_meta=true` only for rhel? We can do as a followup.

jschwinger233 replied:
for now, I think we'd better make this a separate commit with everything we've known about?

ldelossa replied:
Im thinking we can remove the use_meta when we deprecate support for RHEL 8.6. I dont think the original ipsec code which used that flag tried to detect whether we were on RHEL. 

Its handles gracefully in bpf_host regardless if mark or meta is used.

ldelossa committed:
ipsec: add VinE related documentation

Add documentation outlining the move to VinE in v1.18 forward.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

pchaigno commented on daemon/cmd/state.go:
```diff
@@ -318,7 +318,26 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
 		}
```
*Commenting on commit "daemon: Reload bpf_host first in case of IPsec upgrade".*

I think the commit description needs an update. Or alternatively you could simply say you're reapplying commit xxxx and why. By the "As explained by the previous commit" might be confusing when this is the first commit :sweat_smile: Plus, I made a typo back then ("this is/isn't quite enough").

> Nevertheless, this separate regeneration is only performed when we actually need: for IPsec with EKS or AKS IPAM mode.

That part also doesn't match the code anymore.

ldelossa replied:
Yup, let me rewrite that commit, that slipped from my attention.

ldelossa replied:
Done.

pchaigno commented on daemon/cmd/state.go:
```diff
@@ -318,7 +318,26 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
 	}
 
+	if option.Config.EnableIPSec {
+		for _, ep := range state.restored {
```
In the original commit adding this code, we had a big comment here. I think it would be useful to have, if only to mention this logic can be removed again in version XX.

ldelossa replied:
Done

pchaigno commented on bpf/lib/overloadable_skb.h:
```diff
@@ -247,6 +247,14 @@ static __always_inline bool ctx_is_overlay(const struct __sk_buff *ctx)
+static __always_inline bool ctx_is_overlay_encrypted(const struct __sk_buff *ctx)
+{
+	if (!is_defined(HAVE_ENCAP))
+		return false;
```
The call to this function is already guarded with `TUNNEL_MODE` (which is more restrictive than `HAVE_ENCAP` AFAIK) so is this really needed?

ldelossa replied:
It's following the pattern of `ctx_is_overlay`. 

Maybe it's better to be consistent here, tho this is not strictly required.

I'm opting to leave it as is, especially since this code should go away in v1.19. 
If it's a big rub tho, I can remove. 

Going to resolve for now, but feel free to unresolve and @ me if you feel this should definitely change.

pchaigno replied:
Ok, makes sense :+1:

pchaigno commented on bpf/lib/common.h:
```diff
@@ -707,6 +707,13 @@ enum metric_dir {
+ *
+ * the SPI bit can be reused since this magic mark is only used POST encryption.
+ */
+#define MARK_MAGIC_OVERLAY_ENCRYPTED	0x1400
```
In the following commit, you have an implicit assumption that `MARK_MAGIC_OVERLAY_ENCRYPTED = MARK_MAGIC_OVERLAY | something` (due to the imbrication of if conditions). Could we make that explicit here by defining `MARK_MAGIC_OVERLAY_ENCRYPTED` based on `MARK_MAGIC_OVERLAY`?

ldelossa replied:
`imbrication` - now added to my vocabulary. That's a useful one. 

Yup that sounds fine to me.

ldelossa replied:
Done.

pchaigno replied:
> `imbrication` - now added to my vocabulary. That's a useful one.

One of these common French words that exists in English but is rarely used. Basically, How to Sound Posh 101 :sweat_smile:

pchaigno commented on bpf/lib/encrypt.h:
```diff
@@ -222,6 +223,159 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+
+	if (!ipsec_redirect_sec_id_ok(src_sec_identity, dst->sec_identity,
+				      ip_proto))
+		return CTX_ACT_OK;
```
Is this the first time we're tying the encryption to the security identities? I'm a bit worried about the security implications :thinking: I guess both informations come from the same place at least (the ipcache).

Why is this check needed?

ldelossa replied:
This syncs us up to the Wireguard way of handling encryption logic. 
We want to begin to consolidate between the two methods if possible.
AFAICT yeah, info comes from the same place so I originally did not see a problem.

This also puts us in a future place to rely on just the node map's SPI value. 

Additionally, we need to begin checking `src` due to the move to the egress device. 
Because we are now applying encryption policies at the egress device, we loose the "assumption" that the traffic's source is from a pod. We had this assumption prior to moving the IPsec hook to the native device, since it always occurred on pod egress. 

It seemed nice to basically perform the same 'method' of applying IPsec in both scenarios.

pchaigno replied:
:+1:

pchaigno commented on bpf/lib/encrypt.h:
```diff
@@ -222,6 +223,159 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+		}
+		break;
+#endif /* TUNNEL_MODE */
+# endif /* ENABLE_IPv6 */
```
I'm not sure I understand why we need this IPv6 + tunneling case. We don't support IPv6 as an underlay, so in tunneling mode you shouldn't see any pod-to-pod IPv6 traffic passing here. It should all be in the IPv4 case. Of course, you can have non-VXLAN IPv6 traffic, but it won't be pod-to-pod.

ldelossa replied:
The macro is an `ifndef` not `ifdef`. 
We encrypt IPv6 traffic in native routing mode, right?

pchaigno replied:
:facepalm: Missed the `n`. Ok, so makes complete sense :sweat_smile:

pchaigno commented on bpf/lib/encrypt.h:
```diff
@@ -222,6 +223,159 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+		 */
+		if (ctx_is_overlay(ctx)) {
+			/* if this is already encrypted, just pass it to stack */
+			if (ctx_is_overlay_encrypted(ctx))
```
Could we move these two checks outside the IPv4/IPv6 switch? IIUC, we can return early if we're in tunneling mode (`TUNNEL_MODE`) and the packet is not an overlay one or it is but already encrypted:
```c
# if defined(TUNNEL_MODE)
if (!ctx_is_overlay(ctx) || ctx_is_overlay_encrypted(ctx))
        return CTX_ACT_OK;
# endif /* TUNNEL_MODE */
```

That would also simplify my life later on, when I'll need to handle encryption in the IPv6-underlay case.

ldelossa replied:
> we can return early if we're in tunneling mode (TUNNEL_MODE) and the packet is not an overlay one

I don't think this is the case. We'd still want to evaluate the packet to see if its pod-to-pod traffic and not leak the traffic. It may get there due to odd circumstances (like upgrade/downgrade) but we can still encrypt it if its pod2pod right?

We can however, do a very early short circuit by doing
```c
# if defined(TUNNEL_MODE)
if (ctx_is_overlay_encrypted(ctx))
        return CTX_ACT_OK;
# endif 
```

Above and outside of the protocol switch

pchaigno replied:
>  I don't think this is the case. We'd still want to evaluate the packet to see if its pod-to-pod traffic and not leak the traffic. It may get there due to odd circumstances (like upgrade/downgrade) but we can still encrypt it if its pod2pod right?

You're right. I think I was still confused on the `ifndef` below when I wrote this.



> We can however, do a very early short circuit by doing
> 
> ```c
> # if defined(TUNNEL_MODE)
> if (ctx_is_overlay_encrypted(ctx))
>         return CTX_ACT_OK;
> # endif 
> ```
> 
> Above and outside of the protocol switch

I'll take that! :smiley:

ldelossa replied:
implemented.

pchaigno commented on .github/workflows/conformance-clustermesh.yaml:
```diff
@@ -574,11 +574,44 @@ jobs:
+            --junit-file "cilium-junits/${{ env.job_name }} (${{ join(matrix.*, ', ') }}).xml" \
+            --junit-property github_job_step="Run connectivity test (${{ join(matrix.*, ', ') }})" \
+            --test 'no-unexpected-packet-drops'
+          fi
```
Why do we need all this? Isn't the existing check in the connectivity tests enough?

ldelossa replied:
This is a work around for node ID issue we ran into while testing. @jschwinger233 implemented a work around fix and there's an issue to track it. 

I think we will follow up on this one, and it may go away, in the near future.

pchaigno replied:
Would be good to document, in the code or the commit, what's going on here. Temporary fixes tend to stick around and it will likely confuse other reviewers.

thorn3r replied:
I agree, if I was looking at this workflow with no context I'd be confused by this.

ldelossa replied:
Ah! My bad I actually didn't realize that commit had no description. That was shuffled between me and Gray and I totally missed that it was blank. Will update, and I'll add a comment for good measure.

ldelossa replied:
fixed

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -21,6 +21,22 @@ Packets are not encrypted when they are destined to the same node from which
+for overlay encapsulation prior to IPsec encryption when tunnel mode is enabled.
+
+This results in endpoint security IDs which are carried via the VXLan header
+being encrypted on the wire. This is a security benefit.
```
A double passive voice :grimacing: How about:
```
With this change, the security identities used for policy enforcement are encrypted on the wire. 
```

ldelossa replied:
All docs related comments fixed.

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -21,6 +21,22 @@ Packets are not encrypted when they are destined to the same node from which
+v1.18 Encrypted Overlay
+=========================
+Prior to v1.18 IPsec encryption was performed before tunnel encapsulation.
+From Cilium v1.18 and forward Cilium's IPsec encryption datapath will send traffic
```
Some commas to help non-native speakers parse sentences:
"Prior to v1.18,"
"From Cilium v1.18 and forward,"

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -21,6 +21,22 @@ Packets are not encrypted when they are destined to the same node from which
+patching v1.17 to its latest version. Migration specific code was added to
+newer v1.17 releases to support a disruption-less upgrade to v1.18.
+
+Once patched to the newest v1.17 stable release a normal upgrade to v1.18 can
```
One more comma:
"stable release,"

pchaigno reviewed (commented):
Some questions. Not sure there's anything really blocking.

Thanks for the easy-to-review commits btw!

thorn3r reviewed (approved):
looks good for my codeowners

ldelossa commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -21,6 +21,23 @@ Packets are not encrypted when they are destined to the same node from which
 they were sent. This behavior is intended. Encryption would provide no benefits
 in that case, given that the raw traffic can be observed on the node anyway.
 
+v1.18 Encrypted Overlay
```
I think that is correct. Its v1.18 this is introduced in and will just be normal moving forward.

ldelossa commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -21,6 +21,23 @@ Packets are not encrypted when they are destined to the same node from which
+patching v1.17 to its latest version. Migration specific code was added to
+newer v1.17 releases to support a disruption-less upgrade to v1.18.
+
+Once patched to the newest v1.17 stable release, a normal upgrade to v1.18 can
+be performed.
```
Yup, I looked over this document just now. 

Its pretty busy already but I will add a call out or note saying "If upgrading to v1.17 -> v1.18 and using IPsec please read the caveats in {link to documentation section}"

ldelossa replied:
@lambdanis I added a note in the upgrade documentation.

ldelossa commented on Documentation/operations/upgrade.rst:
```diff
@@ -17,6 +17,11 @@ questions, feel free to ping us on `Cilium Slack`_.
 
 .. include:: upgrade-warning.rst
 
+.. note::
```
Done :)

nathanjsweet reviewed (approved):
The regenerate logic makes sense.

jschwinger233 commented on bpf/lib/encap.h:
```diff
@@ -79,12 +79,6 @@ encap_and_redirect_with_nodeid(struct __ctx_buff *ctx, __be32 tunnel_endpoint,
-
 	return __encap_and_redirect_with_nodeid(ctx, tunnel_endpoint,
 						seclabel, dstid, NOT_VTEP_DST,
 						trace);
```
nonblocking, let's cleanup the unnecessary function nest `encap_and_redirect_with_nodeid` -> `__encap_and_redirect_with_nodeid` and unused parameter `encrypt_key` as a followup.

ldelossa replied:
This is handledin: https://github.com/cilium/cilium/pull/39071

@julianwiedmann correct me if I'm wrong. I believe this PR removes the usage of the encrypt_key now.

jschwinger233 reviewed (approved):
Let's run ci-aks and ci-eks.

tommyp1ckles reviewed (approved):
cli looks good

pchaigno commented on bpf/tests/ipsec_from_lxc_tunnel.c:
```diff
@@ -1,9 +0,0 @@
-
-#define EXPECTED_DEST_MAC mac_two
-
-#include "ipsec_from_lxc_generic.h"
```
@ldelossa Shouldn't this file be removed as well?

ldelossa replied:
Yup, think so. 

Can't find a reference to it. This one slipped.

pchaigno commented on bpf/tests/ipsec_redirect_native.c:
```diff
@@ -0,0 +1,98 @@
+	 * The tunnel_endpoint is used in the IPsec hook under test below to
+	 * find the associated NodeID for an egress packet
+	 */
+	ipcache_v4_add_entry(DST_IP, 0, 0xAC, DST_IP, TARGET_SPI);
```
Hm, shouldn't all of this be in a `SETUP` instead of `CHECK`?

pchaigno replied:
Oh, I see the comment above :grimacing: Yeah... we should debug.

ldelossa replied:
I often run into strange anomalies when writing BPF unit tests. I tend to follow the path of least resistance once I find something that works. 

Can't hurt to debug further tho.

pchaigno replied:
I think part of the issue is that we keep calling them BPF unit tests when they're actually integration tests. The SETUP, CHECK, PKTGEN things are for integration tests, but you tried to use them to test just one function (so unit test). It took me a bit to understand that's why you had issues with SETUP.

ldelossa replied:
Oh, I see. Cuz the first example in the BPF testing docs is a unit test which tests one function :-D

pchaigno replied:
Hm, I see. And the docs don't say when to use which.

pchaigno commented on bpf/tests/ipsec_redirect_native.c:
```diff
@@ -0,0 +1,98 @@
+	int i;
+
+	for (i = 0; i < 6; i++)
+		assert(l2->h_dest[i] = SOURCE_MAC[i]);
```
Missing equal sign here.

pchaigno replied:
Also, this is incorrect I believe. The destination L2 address is rewritten by `ipsec_maybe_redirect_to_encrypt` to be `CILIUM_NET_MAC`, not the source L2 address. Anyway, I'm rewriting all this to match the usual integration tests we have for BPF :slightly_smiling_face:

ldelossa replied:
Yikes, that's a bad one 🤦 

> Anyway, I'm rewriting all this to match the usual integration tests we have for BPF 

Sounds good.
